### PR TITLE
Fixes for extra balances, scenarios

### PR DESCRIPTION
**Pull request by antonina-cherednichenko**
This pr contains the following changes:
1. Allow TokenBalance constraint to reduce the balance of an account if needed
2. Add address of USDC(base token) whale for sourcing tokens
3. ~~Mainnet fork~~
4. scenario fixes

### CODE REVIEW

**Comment by antonina-cherednichenko** [comment]
Need to be updated after Jared's pr is merged in.

File: scenario/constraints/TokenBalanceConstraint.ts [hunk 1]
```diff
@@ -29,14 +29,22 @@ export class TokenBalanceConstraint<T extends CometContext, R extends Requiremen
...
             const actor = await getActorAddressFromName(actorName, context);
             const amount: ComparativeAmount = actorsByAsset[assetName][actorName];
             const balance = await asset.balanceOf(actor);
             const decimals = await asset.token.decimals();
             const toTransfer = getToTransferAmount(amount, balance, decimals);
-            await context.sourceTokens(world, toTransfer, asset.address, actor);
+            if (toTransfer < 0) {
+              // Send extra balance to the other address
+              const leftoverHolderAddress = '0x0000000000000000000000000000000000000001';
+              await world.hre.network.provider.send('hardhat_setNextBlockBaseFeePerGas', ['0x0']);
```
Comment (jflatow) [inline scenario/constraints/TokenBalanceConstraint.ts:43]: ⛑️  added a nice helper for this in `context`

File: scenario/constraints/TokenBalanceConstraint.ts [hunk 2]
```diff
@@ -29,14 +29,22 @@ export class TokenBalanceConstraint<T extends CometContext, R extends Requiremen
...
+            const cometActor = context.actors[actorName];
             const actor = await getActorAddressFromName(actorName, context);
             const amount: ComparativeAmount = actorsByAsset[assetName][actorName];
             const balance = await asset.balanceOf(actor);
             const decimals = await asset.token.decimals();
             const toTransfer = getToTransferAmount(amount, balance, decimals);
-            await context.sourceTokens(world, toTransfer, asset.address, actor);
+            if (toTransfer < 0) {
+              // Send extra balance to the other address
+              const leftoverHolderAddress = '0x0000000000000000000000000000000000000001';
```
Comment (jflatow) [inline scenario/constraints/TokenBalanceConstraint.ts:42]: Maybe a const at the top of the file?

**Comment by jflatow** [comment]
> This pr contains the following changes:
> 
> 1. Allow TokenBalance constraint to reduce the balance of an account if needed
> 2. Add address of USDC(base token) whale for sourcing tokens
> 3. ~Mainnet fork~
> 4. scenario fixes

I think you also got rid of the USDC whale here (which is good)?

File: scenario/constraints/TokenBalanceConstraint.ts (hunk 2 above)
Comment (jflatow) [reply scenario/constraints/TokenBalanceConstraint.ts:42]: BTW, there *is* `removeTokens` in TokenSourcer, should we be triggering that?

File: scenario/context/CometContext.ts [hunk 3]
```diff
@@ -271,7 +272,7 @@ async function buildActor(name: string, signer: SignerWithAddress, context: Come
 const getInitialContext = async (world: World): Promise<CometContext> => {
   let deploymentManager = new DeploymentManager(world.base.name, world.hre, { debug: true });
 
-  if (!world.isRemoteFork()) {
+  if (!world.isRemoteFork() || world.base.deployNewComet) {
```
Comment (kevincheng96) [inline scenario/context/CometContext.ts:275]: Do we need this option? Jared's changes will automatically deploy a new Comet for instances where a `roots.json` is either empty or incomplete.

File: scenario/constraints/TokenBalanceConstraint.ts (hunk 2 above)
Comment (kevincheng96) [reply scenario/constraints/TokenBalanceConstraint.ts:42]: Yeah, TokenSourcer *should* be handling negative transfers, so maybe we could move that fix there.

Here's the code in TokenSourcer for reference (not sure why it's not working properly): https://github.com/compound-finance/comet/blob/main/plugins/scenario/utils/TokenSourcer.ts#L26-L27

File: plugins/scenario/World.ts [hunk 4]
```diff
@@ -7,6 +7,8 @@ export type ForkSpec = {
   blockNumber?: number;
   allocation?: number;
   chainId?: number;
+  baseTokenFauceteer?: string;
```
Comment (kevincheng96) [inline plugins/scenario/World.ts:10]: Is this where you set the `USDC whale` address?

**Review by kevincheng96 — COMMENTED** [review]
Nice fix for scenarios 👍

File: plugins/scenario/World.ts (hunk 4 above)
Comment (antonina-cherednichenko) [reply plugins/scenario/World.ts:10]: yes

Comment (jflatow) [reply plugins/scenario/World.ts:10]: not sure about this

Comment (antonina-cherednichenko) [reply plugins/scenario/World.ts:10]: why? it's a way to set whale account for further impersonating. We won't have faucet contracts on the mainnet.

Comment (jflatow) [reply plugins/scenario/World.ts:10]: So token sourcer isn't finding accounts with USDC?

**Comment by jflatow** [comment]
We applied the minimal patch (balance constraints) we needed to #501 and have mainnet scens passing there, so closing

**Pull request closed without merging**